crates/ide_completion/src/item.rs

```diff
@@ -551,9 +569,24 @@ mod tests {
             vec![CompletionRelevance::default()],
             vec![
                 CompletionRelevance { exact_name_match: true, ..CompletionRelevance::default() },
-                CompletionRelevance { exact_type_match: true, ..CompletionRelevance::default() },
+                CompletionRelevance { is_local: true, ..CompletionRelevance::default() },
             ],
-            vec![CompletionRelevance { exact_name_match: true, exact_type_match: true }],
+            vec![CompletionRelevance {
+                exact_name_match: true,
+                is_local: true,
+                ..CompletionRelevance::default()
+            }],
+            vec![CompletionRelevance { exact_type_match: true, ..CompletionRelevance::default() }],
+            vec![CompletionRelevance {
+                exact_name_match: true,
+                exact_type_match: true,
+                ..CompletionRelevance::default()
+            }],
+            vec![CompletionRelevance {
+                exact_name_match: true,
+                exact_type_match: true,
+                is_local: true,
+            }],
```

Comment:
Btw, that's an example of some test-updating busy-work `Expect` saves you from which I've been talking about in that other PR.

Reply:
Hmm, I understand in the other PR, but in this case I'm not sure how best to use `Expect`. The list I modified here is both the test input and the expected output.

Since we don't necessary care about all possible values of `CompletionRelevance` (there are currently `2^3` but there will eventually be many more), this list describes all the instances of `CompletionRelevance` that we actually want to specify the relative ranking of. This list also specifies the desired relative ranking. 

An alternative would be to split those two things up. There would be one input list of `CompletionRelevance` that we care about, and an `Expect` containing the sorted output. But then the input list would be unsorted, which seems like it would be bad for test maintenance in its own way (but perhaps this is a lesser of two evils?). 

Another option might be to generate all possible `CompletionRelevance` and then have an `Expect` check against the sorted output. But that would become overwhelming as we add even a few more `bool` values, and even a single integer or enum with a handful of variants would immediately create too much output.

Reply:
Ah, right, I missed the bit about the input. I generally try to write tests in such a way that input is always a string fixture with Rust code. This doesn’t allow unit-testing internal API details, but, if it’s hard to exercise an api via some user-visible feature, then I think it doesn't matter that much. 

In this particular case, my hypothesis is that 90% of the testing value for relevances we get from all other completion tests that sort the items by relevance. Ie, if we suddenly stop ordering the locals before modules, we'll notice that the order of items in tests. 

That being said, I think we don't actually do sorting by relevance in 

https://github.com/rust-analyzer/rust-analyzer/blob/de360275416ca095102f2b17d6ca1de3bd091fdb/crates/ide_completion/src/lib.rs#L104-L136

?

Seems like we should do, to make our test suite check relative relevance as well.

Reply:
> That being said, I think we don't actually do sorting by relevance in

My thinking here is that determining the proper values in the `CompletionRelevance` struct (whether `exact_name_match` or `is_local` are true/false, for example) is a separate concern from turning that `CompletionRelevance` struct into a score. Following from that, I thought there should be a set of tests which ensure we properly set `CompletionRelevance` (this is an objective true/false), and a different set of tests which check whether the way we turn that `CompletionRelevance` into a score results in a *reasonable* sorting (this is subjective).

I imagine in the future, in order to fix sorting "bugs", we will have to make one of two possible changes:

1. Fix the way some value in `CompletionRelevance` is being set (for example maybe we are setting `is_local` on something we shouldn't be)
2. Fix the way the `CompletionRelevance::score` method computes a score (for example perhaps we accidentally over-weight two obscure things in a way that means that an item with those two things always ends up at the top even though we think it should not)

In the current test design, changes of type 1 would result in one set of tests changing, while changes of type 2 would result in a non-overlapping different set of tests changing (right now changing the scoring algorithm would change only two tests I think). That seems appropriate to me given I believe these are orthogonal behaviors.

Reply:
It's true that these are orthogonal concerns in code as written. However, I find it useful to avoid organizing tests around the specific current separation of concerns. When you refactor code, often times you want to separate concerns a little bit differently the next time around and, if the shape of the tests duplicates the shape of the code, that means you need to change both the code and the tests. 

To give a hypothetical example, lets say that some years down the road we decide that manually computing and ordering completion scores is not cool and instead we just through all the completions into the neural network to sort the order out. 
If all our tests are based around "ensure that this item gets this specific strucured score" and "ensure that structured scores are ordered in this way", we can't actually re-use them. We no longer have structured scores. 

If our tests are based on "make sure that locals sort before modules", then all our tests just work!

The rule-of-thumb here is that, generally, tests should test user-visible features (like the relative ordering of completion items) and should be oblivious of the specific code used to implement them.

Reply:
> The rule-of-thumb here is that, generally, tests should test user-visible features (like the relative ordering of completion items) and should be oblivious of the specific code used to implement them.

Makes sense. There is always a trade-off between testing only at the public API vs testing at some smaller unit, but perhaps with auto-updating `expect` tests the cost of having lots of tests change for a scoring update is pretty low. I can try to clean this up at some point.

Reply:
>but perhaps with auto-updating expect tests the cost of having lots of tests change for a scoring update is pretty low.

Interesting, I didn't quite realize before that expect tests are a vital ingredient. Yeah, when we *want* to change user-visible results, with unfactored tests we may need to update many tests, which is a chore. With expect, that is automated, and you can access the actual impact of the change.